### DISCUSSION TITLE
Problem with HashMaps and Resources

### DISCUSSION THREAD

**Discussion by IsaacGaches** [bevyengine/bevy#6818] [Q&A] [answered]
I'm working on a Minecraft-like chunk system where chunks load and unload around the player. I have a hash map resource that uses a position as a key and the chunk entity as the value. I'm not experienced with bevy or rust so I'm not sure why but, when I try to iterate over the dictionary and remove items, I get errors. Help please.
I've tried a few things but I don't know what I'm doing.

![image](https://user-images.githubusercontent.com/92642398/205150013-699775e0-eb1a-4fc7-9bd8-57995e93d165.png)
![image](https://user-images.githubusercontent.com/92642398/205150683-a3de153d-c871-4360-9cbb-95b507c54e65.png)
![image](https://user-images.githubusercontent.com/92642398/205150555-7acc37a4-c4f3-490a-bbad-a576c0f50c31.png)

**Comment by james7132** [comment, marked as answer]
`Hash` requires `Eq` to be implemented on the type. `Vec2` does not implement it since `f32` does not implement it since `NaN` exists and breaks the contract of `Eq`. If you're doing a chunk map, you could use `UVec2`, which uses integers instead of f32, and it implements both `Eq` and `Hash`.

**Reply by IsaacGaches** [reply]
Thank you very much, that fixed one problem however I still have an insue...
![image](https://user-images.githubusercontent.com/92642398/205165641-6de1c4a9-2551-45cc-ab43-7a4729a162f1.png)

or 

![image](https://user-images.githubusercontent.com/92642398/205166264-409c5932-9a92-4b4f-8938-e31cf59caf8a.png)
![image](https://user-images.githubusercontent.com/92642398/205166311-ef93d5d3-684b-4bb1-8989-3044f66d43dd.png)

**Reply by james7132** [reply]
You probably want [`HashMap::iter_mut`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.iter_mut) instead. `iter` returns immutable references.

**Reply by IsaacGaches** [reply]
![image](https://user-images.githubusercontent.com/92642398/205235411-a1a53228-aa53-4e32-ae20-ed129a842394.png)

**Reply by james7132** [reply]
Ah you're trying to remove the ones that are away. You cannot mutate the `HashMap` structurally while iterating over it. Try [`HashMap::retain`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.retain) instead to keep only the chunks that match a given condition.
